### PR TITLE
[addons] add all needed addon window interface calls

### PR DESCRIPTION
**Pull request by AlwinEsch**
## Description

Next step for binary addon interface rework.

This is nearly the last change on addon Window class. There becomes
all parts added to have needed functionality, also is it with supported
parts now nearly equal with a Python addon window interface.

As last changes on window part comes on next request only the support
for the various control types.

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: xbmc/addons/interfaces/GUI/Window.cpp [hunk 1]
```diff
@@ -280,6 +315,578 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
...
+    CLog::Log(LOGERROR, "Interface_GUIWindow::%s - invalid handler data (kodiBase='%p', handle='%p', key='%p') on addon '%s'",
+                          __FUNCTION__, addon, pAddonWindow, key, addon ? addon->ID().c_str() : "unknown");
+    return -1;
+  }
+
+  CGUIWindow *pWindow = dynamic_cast<CGUIWindow*>(g_windowManager.GetWindow(pAddonWindow->m_windowId));
+  if (!pWindow)
+    return -1;
+
+
```
Comment (ksooo) [inline xbmc/addons/interfaces/GUI/Window.cpp]: Could you please remove one of the two empty lines here and everywhere else in this source file. Seen thin in several places from here.

File: xbmc/addons/interfaces/GUI/Window.cpp [hunk 2]
```diff
@@ -280,6 +315,578 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
...
+  }
+
+  CGUIWindow *pWindow = dynamic_cast<CGUIWindow*>(g_windowManager.GetWindow(pAddonWindow->m_windowId));
+  if (!pWindow)
+    return;
+
+
+  std::string lowerKey = key;
+  StringUtils::ToLower(lowerKey);
+  Interface_GUIGeneral::lock();
```
Comment (ksooo) [inline xbmc/addons/interfaces/GUI/Window.cpp]: To follow your own style you should insert an empty line before this call (after the tolower call). :-)

File: xbmc/addons/interfaces/GUI/Window.cpp [hunk 3]
```diff
@@ -280,6 +315,578 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
...
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  CGUIAddonWindow* pAddonWindow = static_cast<CGUIAddonWindow*>(handle);
+  if (!addon || !pAddonWindow)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIWindow::%s - invalid handler data (kodiBase='%p', handle='%p') on addon '%s'",
+                          __FUNCTION__, addon, pAddonWindow, addon ? addon->ID().c_str() : "unknown");
+    return false;
+  }
+
+  CGUIWindow *pWindow = dynamic_cast<CGUIWindow*>(g_windowManager.GetWindow(pAddonWindow->m_windowId));
```
Comment (ksooo) [inline xbmc/addons/interfaces/GUI/Window.cpp]: If pAddonWindow is of type <code>CGUIAddonWindow</code> which is guaranteed if you come across this code line, it is save that pAddonWindow is a <code>CGUIWindow</code>, because <code>CGUIAddonWindow</code> is derived from <code>CGUIMediaWindow</code> which is derived from <code>CGUIWindow</code>, right? You don't need pWindow here and everywhere in the code you introduced in this PR. You can call directly on pAddonWindow. ;-)

File: xbmc/addons/interfaces/GUI/Window.cpp [hunk 4]
```diff
@@ -280,6 +315,578 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
...
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  CGUIAddonWindow* pAddonWindow = static_cast<CGUIAddonWindow*>(handle);
+  if (!addon || !pAddonWindow)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIWindow::%s - invalid handler data (kodiBase='%p', handle='%p') on addon '%s'",
+                          __FUNCTION__, addon, handle, addon ? addon->ID().c_str() : "unknown");
+    return;
+  }
+
+  pAddonWindow->MarkDirtyRegion();
```
Comment (ksooo) [inline xbmc/addons/interfaces/GUI/Window.cpp]: Again, why no lock/unlock before/after this call?

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/definitions.h [hunk 5]
```diff
@@ -175,9 +176,39 @@ typedef struct AddonToKodiFuncTable_kodi_gui_window
...
+  void (*remove_list_item)(void* kodiBase, void* handle, void* item);
+  void* (*get_list_item)(void* kodiBase, void* handle, int list_position);
+  void (*set_current_list_position)(void* kodiBase, void* handle, int list_position);
+  int (*get_current_list_position)(void* kodiBase, void* handle);
+  int (*get_list_size)(void* kodiBase, void* handle);
+  void (*set_container_property)(void* kodiBase, void* handle, const char* key, const char* value);
+  void (*set_container_content)(void* kodiBase, void* handle, const char* value);
+  int (*get_current_container_id)(void* kodiBase, void* handle);
+
+  /* Various calls */
```
Comment (ksooo) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/definitions.h]: minor nitpicking: these are not calls, but functions. Seen multiple times in this source file.

**Review by ksooo — CHANGES_REQUESTED** [review]
As  usual I have some small questions and suggestions.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/definitions.h (hunk 5 above)
Comment (AlwinEsch) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/definitions.h]: Upps, had forgotten the place

File: xbmc/addons/interfaces/GUI/Window.cpp [hunk 6]
```diff
@@ -280,6 +315,578 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
...
+    return -1;
+
+  Interface_GUIGeneral::lock();
+  int control_id = pWindow->GetFocusedControlID();
+  Interface_GUIGeneral::unlock();
+
+  if (control_id == -1)
+    CLog::Log(LOGERROR, "Interface_GUIWindow - %s: %s - No control in this window has focus", __FUNCTION__, addon->Name().c_str());
+
+  return control_id;
```
Comment (AlwinEsch) [inline xbmc/addons/interfaces/GUI/Window.cpp]: Upss, was a dead corpse.

File: xbmc/addons/interfaces/GUI/Window.cpp [hunk 7]
```diff
@@ -280,6 +315,578 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
...
+
+  if (!pWindow->GetControl(control_id))
+  {
+    CLog::Log(LOGERROR, "Interface_GUIWindow - %s: %s - Control does not exist in window", __FUNCTION__, addon->Name().c_str());
+    return false;
+  }
+
+  Interface_GUIGeneral::lock();
+  CGUIMessage msg = CGUIMessage(GUI_MSG_SETFOCUS, pAddonWindow->m_windowId, control_id);
+  pWindow->OnMessage(msg);
```
Comment (AlwinEsch) [inline xbmc/addons/interfaces/GUI/Window.cpp]: Have testet, compiler does not accept to have in OnMessage, I change it to 
```cpp
  CGUIMessage msg(GUI_MSG_SETFOCUS, pAddonWindow->m_windowId, control_id);
  pWindow->OnMessage(msg);
```
to match the others.

File: xbmc/addons/interfaces/GUI/Window.cpp (hunk 3 above)
Comment (AlwinEsch) [inline xbmc/addons/interfaces/GUI/Window.cpp]: The path is a remnant of the beginning

https://github.com/xbmc/xbmc/blob/4e9cb43928abbddd5d683cb5d9d66c72b3bbf88e/xbmc/addons/AddonCallbacksGUI.cpp#L388_L391

I think is to confirm Window registered with `g_windowManager.Add(window);`.

But from my side it can also be removed.

**Comment by AlwinEsch** [comment]
Is updated

File: xbmc/addons/interfaces/GUI/Window.cpp (hunk 6 above)
Comment (ksooo) [inline xbmc/addons/interfaces/GUI/Window.cpp]: Okay.

**Review by ksooo — APPROVED** [review]

**Comment by AlwinEsch** [comment]
jenkins build this please

**Comment by AchimTuran** [comment]
@AlwinEsch 
I reviewed your PR. If you have further questions or suggestions please ping on Github, Slack... 

Only minor issues from my side and it is up to you to accept or reject them.

Again thank you for your work, I'm really looking forward to use it in my upcoming AudioDSP binary add-on dialogs.

Keep up the good work.

**Comment by ksooo** [comment]
jenkins build and merge

File: xbmc/addons/interfaces/GUI/Window.cpp [hunk 8]
```diff
@@ -316,11 +625,50 @@ void Interface_GUIWindow::add_list_item(void* kodiBase, void* handle, void* item
   }
 
   Interface_GUIGeneral::lock();
-  pAddonWindow->AddItem(pItem, item_position);
+  pAddonWindow->AddItem(pItem, list_position);
```
Comment (AlwinEsch) [inline xbmc/addons/interfaces/GUI/Window.cpp]: The use as "list_position" is to match the for "C" standardized string format. The other is normally frowned upon for "C".

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/Window.h [hunk 9]
```diff
@@ -166,6 +166,232 @@ namespace gui
...
+    ///
+    ///
+    std::string GetProperty(const std::string& key) const
+    {
+      std::string label;
+      char* ret = m_interface->kodi_gui->window->get_property(m_interface->kodiBase, m_controlHandle, key.c_str());
+      if (ret != nullptr)
+      {
+        if (std::strlen(ret))
+          label = ret;
```
Comment (AlwinEsch) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/Window.h]: Is a bit to prevent the C++ class construction and copy from the not present string, not complete sure what becomes done in compiled code, but thought why let C++ work if C nothing have.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/definitions.h [hunk 10]
```diff
@@ -175,9 +176,39 @@ typedef struct AddonToKodiFuncTable_kodi_gui_window
...
+  bool (*get_property_bool)(void* kodiBase, void* handle, const char* key);
+  double (*get_property_double)(void* kodiBase, void* handle, const char* key);
+  void (*clear_properties)(void* kodiBase, void* handle);
+  void (*clear_property)(void* kodiBase, void* handle, const char* key);
+
+  /* List item functions */
   void (*clear_item_list)(void* kodiBase, void* handle);
-  void (*add_list_item)(void* kodiBase, void* handle, void* item, int item_position);
-  void* (*get_list_item)(void* kodiBase, void* handle, int listPos);
+  void (*add_list_item)(void* kodiBase, void* handle, void* item, int list_position);
```
Comment (AlwinEsch) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/definitions.h]: Has leaved only "kodiBase" in camel case to see for me that is C++ on Kodi related.

File: xbmc/addons/interfaces/GUI/Window.cpp [hunk 11]
```diff
@@ -280,7 +315,281 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
...
+//@}
+
+/*!
+ * Window control functions
+ */
+//@{
+bool Interface_GUIWindow::set_focus_id(void* kodiBase, void* handle, int control_id)
+{
+  CAddonDll* addon = static_cast<CAddonDll*>(kodiBase);
+  CGUIAddonWindow* pAddonWindow = static_cast<CGUIAddonWindow*>(handle);
```
Comment (AlwinEsch) [inline xbmc/addons/interfaces/GUI/Window.cpp]: @ksooo what you thing about `reinterpret_cast`?
For me, this is a final solution if nothing else is possible.

File: xbmc/addons/interfaces/GUI/Window.cpp [hunk 12]
```diff
@@ -280,7 +315,281 @@ bool Interface_GUIWindow::do_modal(void* kodiBase, void* handle)
...
+    CLog::Log(LOGERROR, "Interface_GUIWindow::%s - invalid handler data (kodiBase='%p', handle='%p', key='%p') on addon '%s'",
+                          __FUNCTION__, addon, pAddonWindow, key, addon ? addon->ID().c_str() : "unknown");
+    return;
+  }
+
+  std::string lowerKey = key;
+  StringUtils::ToLower(lowerKey);
+
+  Interface_GUIGeneral::lock();
+  pAddonWindow->SetProperty(lowerKey, "");
```
Comment (AlwinEsch) [inline xbmc/addons/interfaces/GUI/Window.cpp]: This reset the property on given key value.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/Window.h [hunk 13]
```diff
@@ -166,6 +166,232 @@ namespace gui
...
+    /// @param[in] key                  string - property name.
+    /// @param[in] value                string or unicode - value of property.
+    ///
+    /// @note  Key is NOT case sensitive. Setting value to an empty string is
+    ///        equivalent to clearProperty(key).\n
+    ///        You can use the above as keywords for arguments and skip certain
+    ///        optional arguments.\n
+    ///        Once you use a keyword, all following arguments require the keyword.
+    ///
+    void SetProperty(const std::string& key, const std::string& value)
```
Comment (AlwinEsch) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/Window.h]: Thanks! @ksooo has brought me to them.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/Window.h [hunk 14]
```diff
@@ -227,6 +481,135 @@ namespace gui
...
 
     //==========================================================================
+    ///
+    /// \ingroup cpp_kodi_gui_CWindow
+    /// @brief To set position of selected part in list.
+    ///
+    /// @param[in] listPos              Position to use
+    ///
+    ///
+    void SetCurrentListPosition(int listPos)
```
Comment (AlwinEsch) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/Window.h]: I look, maybe yes.

File: xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/Window.h [hunk 15]
```diff
@@ -227,6 +481,135 @@ namespace gui
...
+    //--------------------------------------------------------------------------
+    //@}
+
+    //==========================================================================
+    ///
+    /// \ingroup cpp_kodi_gui_CWindow
+    /// @brief To inform Kodi that it need to render region new.
+    ///
+    ///
+    void MarkDirtyRegion()
```
Comment (AlwinEsch) [inline xbmc/addons/kodi-addon-dev-kit/include/kodi/gui/Window.h]: No Problem :)
With the documentation is currently added what I done some time ago. Currently only present but not direct adjusted for doxygen. Because of the size comes the optimization on separate requests.

File: xbmc/addons/interfaces/GUI/Window.cpp (hunk 11 above)
Comment (ksooo) [inline xbmc/addons/interfaces/GUI/Window.cpp]: Whenever you cast a void to something else, imo, reinterpret_cast is the best choice.

File: xbmc/addons/interfaces/GUI/Window.cpp [hunk 16]
```diff
@@ -348,6 +696,130 @@ void* Interface_GUIWindow::get_list_item(void* kodiBase, void* handle, int listP
...
+  CGUIAddonWindow* pAddonWindow = static_cast<CGUIAddonWindow*>(handle);
+  if (!addon || !pAddonWindow)
+  {
+    CLog::Log(LOGERROR, "Interface_GUIWindow::%s - invalid handler data (kodiBase='%p', handle='%p') on addon '%s'",
+                          __FUNCTION__, addon, handle, addon ? addon->ID().c_str() : "unknown");
+    return;
+  }
+
+  Interface_GUIGeneral::lock();
+  pAddonWindow->SetCurrentListPosition(list_position);
```
Comment (ksooo) [inline xbmc/addons/interfaces/GUI/Window.cpp]: I see it vica versa - for C functions, I would use underline names, for C++ camel case. We're not consistent here in any way.

But may I suggest to do this kind of cleanup in a separate PR, once all stuff from Alwin is in? I think we agreed on this already.
